[FEAT] 닉네임, 성별, 출생년도, 취향 등록 기능 구현

Kim-TaeUk opened the pull request:
## Related Issue
<!-- feat/#issue -> dev와 같이 반영 브랜치를 표시 -->
<!-- closed #issue로 merge되면 issue가 자동으로 close되게 -->
- feat/#24 -> dev
- close #24

## Key Changes
<!-- 최대한 자세히 -->
- 닉네임 검증
  - 이전 뷰에서 닉네임에 대한 검증을 하긴 하지만, 해당 api에서도 중복으로 검증합니다.
  - null, 빈칸(""), 공백으로 시작하거나 끝나는, 길이 조건(2~10자), 정규표현식, 중복 닉네임
  - 중복 닉네임 검증했음에도 DB에 insert 되는 경우에 발생할 수 있는 DataIntegrityViolationException을 잡아 처리하는 handler를 추가했습니다.
  - 닉네임 unique constraint의 이름을 정할 수 없어서 @Table에 새로 정의했습니다.
- 성별 검증
  - null, 빈칸(""), M 또는 F 이외의 String 값
  - Enum이라 검증을 위해 custom annotation, validator 사용했습니다.
- 출생년도 검증
  - null, 1900 ~ 2024 이외의 Integer 값 혹은 Integer가 아닌 값
- 취향 검증
  - null, 프로퍼티의 요소값이 잘못된 경우
  - 사용되는 요소 이름: romanceFantasy, romance, fantasy, modernFantasy, drama, lightNovel, wuxia, mystery, BL

## To Reviewers
<!-- 모호한 점, Key Changes에서 부족한 내용 -->
- GlobalExceptionHandler를 exception/handler에 패키징했습니다.
- 공통 에러 응답에서 code, description 프로퍼티 2개로 내려주고 있는데, spring boot validation 사용한 것을 잡아서 처리할 때 code에 status code와 동일한 내용을 내려주는 것을 개선하고 싶은데 좋은 생각있으면 말해주세요~ (ex. @NotNull validation할 때 http status code로 400 Bad Request를 내려줌에도, body에 "code": "BAD_REQUEST"를 내려주는 것,,!)
- 현재 년도가 2024로 하드 코딩되어 있는데, 어노테이션에서 동적으로 쓰는 방법을 아시나요?

Kim-TaeUk committed:
[FEAT] 닉네임, 성별, 출생년도 등록 엔드포인트 추가

Kim-TaeUk committed:
[FEAT] 닉네임, 성별, 출생년도 등록 DTO 추가

Kim-TaeUk committed:
[FEAT] User 엔티티에 updateUserInfo 메서드 추가

닉네임, 성별, 출생년도 등록

Kim-TaeUk committed:
[FEAT] 닉네임, 성별, 출생년도 등록 서비스 로직 추가

User 엔티티의 updateUserInfo 메서드 사용

Kim-TaeUk committed:
[REFACTOR] User 엔티티에 updateUserInfo 메서드 시그니처 변경

기존 방식에서는 파라미터의 시그니처를 확인하기 위해 User를 참고했어야 함

Kim-TaeUk committed:
[FEAT] enum Gender 검증을 위한 custom annotation 추가

Kim-TaeUk committed:
[FEAT] enum Gender validator 정의

null check, blank check
M/F만 허용

Kim-TaeUk committed:
[FIX] request에서 gender의 type 변경

from Gender(enum) to String

Kim-TaeUk committed:
[FEAT] enum Gender validator에서 사용되는 CustomError 정의

INVALID_GENDER, INVALID_GENDER_NULL, INVALID_GENDER_BLANK

Kim-TaeUk committed:
[FEAT] RegisterUserInfoRequest DTO에서 Gender validation 추가

NotNull, GenderConstraint

Kim-TaeUk committed:
[FIX] User 엔티티의 updateUserInfo 메서드의 필드 바인딩 타입 변경

DTO에서 String -> Gender 변경, Gender.valueOf()로 바인딩

Kim-TaeUk committed:
[FEAT] UserController에서 검증을 위한 @Valid 추가

Kim-TaeUk committed:
[FIX] NicknameConstraint의 Target에 FIELD 추가

record(DTO)의 field에서 사용

Kim-TaeUk committed:
[FEAT] RegisterUserInfoRequest DTO에서 nickname validation 추가

Kim-TaeUk committed:
[FEAT] User 엔티티에 셋하기 전에 nickname 검증 추가

DB에 조회하여 무결성 검증

Kim-TaeUk committed:
[REFACTOR] servie layer의 nickname 검증 로직 메서드로 추출

Kim-TaeUk committed:
[FEAT] ControllerAdvice에 DataIntegrityViolationException handler 추가

nickname으로 인한 데이터 무결성이 깨졌을 때 터지는 exception

Kim-TaeUk committed:
[REFACTOR] handler 패키징

GlobalExceptionHandler를 exception/handler 패키지로

Kim-TaeUk committed:
[FEAT] UserExceptionHandler 추가

DataIntegrityViolationExceptionHandler를 GlobalExceptionHandler에서 분리

Kim-TaeUk committed:
[REMOVE] UserExceptionHandler 삭제

Kim-TaeUk committed:
[REFACTOR] User 엔티티의 nickname unique 설정 방식 변경

@Column에 명시하는 방식에서, @Table에 명시하는 방식을 통해 제약조건의 이름 설정 가능

Kim-TaeUk committed:
[REFACTOR] DataIntegrityViolationExceptionHandler를 GlobalExceptionHandler 안에 포함

rootCauseMessage를 통해 nickname의 unique 조건을 위반했는지 판단 후 분기하여 에러 처리

Kim-TaeUk committed:
[REMOVE] 불필요한 출력문 제거

Kim-TaeUk committed:
[FIX] request에서 birth의 type 변경

from Year to Integer

Kim-TaeUk committed:
[FIX] User 엔티티의 updateUserInfo 메서드의 필드 바인딩 타입 변경

DTO에서 Year -> Integer 변경, Integer.valueOf()로 바인딩

Kim-TaeUk committed:
[FEAT] request dto validation 추가

not null, 1900~2024년 사이

Kim-TaeUk committed:
[RENAME] MethodArgumentNotValidExceptionHandler로 rename

Kim-TaeUk committed:
[FEAT] request DTO에 genrePreferences(선호 장르) 추가

Kim-TaeUk committed:
[FEAT] GenreRepository 추가, findByGenre 메서드 추가

Kim-TaeUk committed:
[FEAT] CustomGenreException 추가

Kim-TaeUk committed:
[FEAT] CustomGenreError 추가

Kim-TaeUk committed:
[FEAT] GenrePreference에 정적 팩터리 create 메서드 추가

Kim-TaeUk committed:
[FEAT] 선호 장르 추가 로직 추가

Kim-TaeUk committed:
[FIX] 선호 장르 validation message 추가

Kim-TaeUk commented on src/main/java/org/websoso/WSSServer/domain/User.java:
```diff
@@ -7,26 +7,34 @@
 @Entity
 @Getter
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Table(uniqueConstraints = {
+        @UniqueConstraint(
+                name = "UNIQUE_NICKNAME_CONSTRAINT",
+                columnNames = "nickname")
+})
```
`@Table`에서 설정함으로써 unique 제약 조건의 이름 설정 가능

기존 `@Column`에서 unique = true로 설정해줄 때는 Hibernate가 임의의 이름으로 설정해버림

Kim-TaeUk commented on src/main/java/org/websoso/WSSServer/dto/user/RegisterUserInfoRequest.java:
```diff
@@ -0,0 +1,25 @@
+        String gender,
+
+        @NotNull(message = "출생연도는 null일 수 없습니다.")
+        @Range(min = 1900, max = 2024, message = "출생연도는 1900 ~ 2024 사이여야 합니다.")
```
현재 연도가 2024로 하드 코딩되어 있는데, 어노테이션에서 동적으로 가져오는 방법을 못 찾아서 일단 이렇게 해두었습니다.

Kim-TaeUk commented on src/main/java/org/websoso/WSSServer/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -18,7 +22,7 @@
 
     @ExceptionHandler(MethodArgumentNotValidException.class)
-    public ResponseEntity<ErrorResult> InvalidRequestBodyExceptionHandler(MethodArgumentNotValidException e) {
+    public ResponseEntity<ErrorResult> MethodArgumentNotValidExceptionHandler(MethodArgumentNotValidException e) {
```
잡아서 처리하는 예외의 이름으로 handler를 네이밍하고 있어서 rename했습니다.

Kim-TaeUk commented on src/main/java/org/websoso/WSSServer/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -34,6 +38,22 @@ public ResponseEntity<ErrorResult> HttpMessageNotReadableExceptionHandler(HttpMe
                 .body(new ErrorResult(BAD_REQUEST.name(), "잘못된 JSON 형식입니다."));
     }
 
+    @ExceptionHandler(DataIntegrityViolationException.class)
+    public ResponseEntity<ErrorResult> DataIntegrityViolationExceptionHandler(DataIntegrityViolationException e) {
+        log.error("[DataIntegrityViolationException] exception ", e);
+
+        String rootCauseMessage = Objects.requireNonNull(e.getRootCause()).getMessage();
+        if (rootCauseMessage != null && rootCauseMessage.contains("UNIQUE_NICKNAME_CONSTRAINT")) {
+            return ResponseEntity
+                    .status(CONFLICT)
+                    .body(new ErrorResult(DUPLICATED_NICKNAME.getCode(), DUPLICATED_NICKNAME.getDescription()));
+        }
+
+        return ResponseEntity
+                .status(CONFLICT)
+                .body(new ErrorResult(CONFLICT.name(), "DB 무결성 제약조건이 위반되었습니다."));
+    }
+
```
DataIntegrityViolationException이 unique 제약조건을 위반할 때뿐만 아니라, DB의 무결성 제약조건을 위반할 때 터지는 예외인데
nickname(unique 제약조건)이 터질 때를 구분하고 싶어서 따로 분리했습니다.

@Table에서 설정해준 제약조건의 이름인 UNIQUE_NICKNAME_CONSTRAINT을 포함하는지 확인하는 방식입니다.
그렇게 좋은 방식은 아닌 것 같은데, 개선점 말씀해주세요~

Kim-TaeUk commented on src/main/java/org/websoso/WSSServer/validation/NicknameConstraint.java:
```diff
@@ -11,7 +12,7 @@
 @Documented
 @Constraint(validatedBy = NicknameValidator.class)
-@Target({PARAMETER})
+@Target({PARAMETER, FIELD})
```
기존에는 controller에서 request param에서만 쓰여서 parameter로만 설정해두었는데,
record(dto)에서 쓰이게 돼서 field도 추가했습니다.

ChaeAg commented on src/main/java/org/websoso/WSSServer/service/UserService.java:
```diff
@@ -86,4 +91,26 @@ public MyProfileResponse getMyProfileInfo(User user) {
+                            .orElseThrow(() -> new CustomGenreException(GENRE_NOT_FOUND, "genre with tge given genreName is not found"));
+                    return GenrePreference.create(user, genre);
+                })
+                .collect(Collectors.toList());
```
P4; 해당 map 구문을 아래와 같이 하는 건 어떨까요? 조금 더 간결해질 수 있을 것 같아 제안합니다!

```java
.map(preferGenreName -> genreRepository.findByGenreName(preferGenreName)
                .map(genre -> GenrePreference.create(user, genre))
                .orElseThrow(() -> new CustomGenreException(GENRE_NOT_FOUND, "genre with the given genreName is not found"))
        )
```

Kim-TaeUk replied:
제안된 방식으로 했을 때는 genreName에 해당하는 Genre를 GenreRepository에서 찾을 때 없는 경우 예외를 못 터뜨리지 않나요?

ChaeAg replied:
optional의 map 은 값이 없다면 orElse 쪽으로 넘어가는 걸로 아는데.. 아닌가요??

Kim-TaeUk replied:
와 그러네요?!?!?!
모르는 부분이었습니다!!! 알려줘서 감사합니다!
반영할게요!

ChaeAg reviewed (approved):
수고하셨습니다!
특별히 수정할 것은 없어보입니다!
본문에 남기신 것은 함께 알아보아요 ^^..

Kim-TaeUk commented on src/main/java/org/websoso/WSSServer/domain/User.java:
```diff
@@ -59,6 +67,12 @@ public void updateProfileStatus(Boolean profileStatus) {
         this.isProfilePublic = profileStatus;
     }
 
+    public void updateUserInfo(RegisterUserInfoRequest registerUserInfoRequest) {
+        this.nickname = registerUserInfoRequest.nickname();
+        this.gender = Gender.valueOf(registerUserInfoRequest.gender());
+        this.birth = Year.of(registerUserInfoRequest.birth());
+    }
```
지금 사용한 케이스에서는 딱히 필요한 것 같지는 않지만,
메서드 파라미터의 개수가 많아지는 경우 dto를 파라미터로 넘기는 방식도 있습니다.

메서드 시그니처로 인해 혼란스러울 가능성이 있는 경우에 사용하면 좋을 것 같네요.
ex. 파라미터가 5개인데 모두 String 타입인 경우, 파라미터의 순서를 알려면 해당 메서드가 정의된 곳을 확인해야 하는 번거로움이 있음. (물론 intelliJ에서는 cmd + P하면 나오긴 함 ㅋㅋ,,)

Kim-TaeUk committed:
[REFACTOR] Optional API map 로직 리팩터링

중간 genre 변수 제거를 통해 가독성 개선

rinarina0429 commented on src/main/java/org/websoso/WSSServer/exception/error/CustomUserError.java:
```diff
@@ -23,7 +23,10 @@ public enum CustomUserError implements ICustomError {
     INVALID_USER_ID("USER-008", "유효하지 않은 ID입니다.", BAD_REQUEST),
     DUPLICATED_NICKNAME("USER-009", "중복된 닉네임입니다.", CONFLICT),
-    INVALID_PROFILE_STATUS("USER-010", "프로필 상태는 이미 설정된 값과 동일합니다.", BAD_REQUEST);
+    INVALID_PROFILE_STATUS("USER-010", "프로필 상태는 이미 설정된 값과 동일합니다.", BAD_REQUEST),
```
p3;
```java
ALREADY_SET_NICKNAME("USER-011", "닉네임은 이미 설정된 닉네임과 동일합니다.", BAD_REQUEST);
```
이것처럼 `INVALID_PROFILE_STATUS`보다는 `ALREADY_SET_PROFILE_STATUS`는 어떠신가요? 뭔가 INVALID라고 하면 기존의 의미와 다르게 전달되는 것 같아서요

Kim-TaeUk replied:
그러네요~~!
반영했습니다! 👍

rinarina0429 commented on src/main/java/org/websoso/WSSServer/service/UserService.java:
```diff
@@ -86,4 +90,27 @@ public MyProfileResponse getMyProfileInfo(User user) {
+                .collect(Collectors.toList());
+
+        genrePreferenceRepository.saveAll(preferGenres);
+    }
```
p2;
아래처럼 함수 분리하는건 어떠신가요? 여기서 바로 분리한거라 제대로 된건지는 확인이 필요하겠지만... 지금은 map 안에 map이 있기도 하고 잘 읽히는 느낌이 아니라서요.
그리고 제가 올려둔 pr#120 에도 GenreService 내에 findGenreOrThrow가 있어서 나중에 합치면서 그 함수 하나로 같이 이용할 수 있을 것 같기도 하구요?
```suggestion
public void registerUserInfo(User user, RegisterUserInfoRequest request) {
    validateNickname(request.nickname());
    user.updateUserInfo(request);

    List<GenrePreference> preferGenres = createGenrePreferences(user, request.genrePreferences());
    genrePreferenceRepository.saveAll(preferGenres);
}

private void validateNickname(String nickname) {
    if (userRepository.existsByNickname(nickname)) {
        throw new CustomUserException(DUPLICATED_NICKNAME, "Nickname is duplicated.");
    }
}

private List<GenrePreference> createGenrePreferences(User user, List<String> genreNames) {
    return genreNames.stream()
            .map(this::findGenreOrThrow)
            .map(genre -> GenrePreference.create(user, genre))
            .collect(Collectors.toList());
}

private Genre findGenreOrThrow(String genreName) {
    return genreRepository.findByGenreName(genreName)
            .orElseThrow(() -> new CustomGenreException(GENRE_NOT_FOUND, 
                    genre with the given genreName is not found));
}
```

Kim-TaeUk replied:
와우 반영했습니다
매우 야무져요!!
메서드 참조 쓰니까 훨씬 가독성이 좋아지네요!

rinarina0429 reviewed (changes requested):
수고하셨습니다~!! 리뷰 한번만 확인 부탁드립니다아

rinarina0429 commented on src/main/java/org/websoso/WSSServer/dto/user/RegisterUserInfoRequest.java:
```diff
@@ -0,0 +1,25 @@
+        String gender,
+
+        @NotNull(message = "출생연도는 null일 수 없습니다.")
+        @Range(min = 1900, max = 2024, message = "출생연도는 1900 ~ 2024 사이여야 합니다.")
```
`@BirthYearConstraint` 같은거 만들어서 달면 되지 않을까요?
```java
String message() default "출생연도는 {min} ~ {max} 사이여야 합니다.";
```
넣어두고 Validator에서
```java
String messageTemplate = context.getDefaultConstraintMessageTemplate();
String dynamicMessage = messageTemplate
        .replace("{min}", String.valueOf(MIN_YEAR))
        .replace("{max}", String.valueOf(Year.now().getValue()));
```
 요롱식으로??

Kim-TaeUk replied:
custom으로 만들지 않고 어노테이션에서 동적으로 쓰고 싶어서 그런건데 ㅎㅎ,,, (방법이 없대요 ㅎㅎ,,)
그냥 custom으로 만들어야겠네요~

rinarina0429 replied:
방법이 없다니 허허...ㅜㅜ커스텀을 쓸 수 밖에 없겠네요
근데 메세지에는 2024로 해두고 해가 바뀔 때마다 수정하는 방식인가요?

Kim-TaeUk replied:
`유효하지 않은 출생년도입니다.`로 수정했습니다!

Kim-TaeUk committed:
[REFACTOR] 닉네임, 성별, 출생년도 등록 - 서비스 레이어 메인 메서드 리팩터링

- findByGenreNameOrThrow 메서드로 추출: GenreName으로 Genre 찾는 로직
- createGenrePreferences 메서드로 추출: 선호 장르 생성 로직

Kim-TaeUk committed:
[RENAME] 이미 설정된 프로필에 대한 CustomUserError rename

INVALID_PROFILE_STATUS -> ALREADY_SET_PROFILE_STATUS

Kim-TaeUk committed:
[FEAT] 현재 년도를 동적으로 판단하는 Custom annotation @BirthConstraint 추가

Kim-TaeUk committed:
[FIX] RegisterUserInfoRequest.birth를 @BirthConstraint로 검증

Kim-TaeUk committed:
[FIX] error response에 내려줄 message 변경

Kim-TaeUk committed:
Merge branch 'dev' into feat/#24
